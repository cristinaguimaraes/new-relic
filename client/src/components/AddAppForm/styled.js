import styled from 'styled-components';

export const Helper = styled.span`
 display: block;
 color: red;
 margin-top: 4px;
 font-family: 'Helvetica Bold';

  &:first-letter {
    text-transform: capitalize;
  }
`;

export const Input = styled.input`
font-family: 'Helvetica';
height: 48px;
min-height: 48px;
`;

export const Label = styled.label`
font-family: 'Helvetica Bold';
margin-top: 15px;
margin-bottom: 5px;
  &:first-letter {
    text-transform: capitalize;
  }
`;

export const Form = styled.form`
  width: 50%;
  display: flex;
  flex-direction: column;
  padding: 24px 16px;
  border-radius: 4px;
  margin-top: 24px;
  border: solid 1px

`;

export const ButtonWrapper = styled.div`
width: 100%;
display:flex;
justify-content:center
`;

export const Button = styled.button`
width: 80px;
margin-top: 15px;
border-radius: 6px;
height: 40px;
`;

export const Title = styled.div`
font-size: 35px;
font-family: 'Helvetica Bold';
color: #000000;
line-height: 43px;
margin-bottom: 15px;
text-align: center
`;

