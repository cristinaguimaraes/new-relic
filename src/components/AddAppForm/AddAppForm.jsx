import React, {useState} from "react";
import {
  Helper,
  Input,
  Label,
  Form,
  Button,
  ButtonWrapper,
  Title
} from "./styled";
import {func} from "prop-types";

const AddAppForm = ({onAddApp, onShowToastMessage}) => {
  const [app, setApp] = useState({
    name: "",
    contributors: [],
    version: "",
    host: [],
    apdex: ""
  });
  const [errors, setErrors] = useState({
    name: "",
    contributors: "",
    version: "",
    host: "",
    apdex: ""
  });

  const handleOnChange = event => {
    const value =
      event.target.name === "contributors" || event.target.name === "host"
        ? event.target.value.split(",")
        : event.target.value;
    setApp({...app, [event.target.name]: value});
    setErrors({...errors, [event.target.name]: ""});
  };

  const handleOnSubmit = event => {
    event.preventDefault();

    const errors = {};
    if (!app.name) errors.name = "Please add the name";

    if (app.contributors.length === 0) {
      errors.contributors = "Please add the contributors";
    }
    if (!app.version) {
      errors.version = "Please add the version";
    }
    if (app.host.length === 0) {
      errors.host = "Please add the host";
    }
    if (!app.apdex) {
      errors.apdex = "Please add apdex";
    }

    if (Object.keys(errors).length === 0) {
      onAddApp({app});
      //If 'onAddApp' was dispatching an action calling the API, the toast should appear just in case of Success. It is here because of the scope of the project.
      onShowToastMessage({messageKey: "App successfully added to host", type: 'success'});
      setApp({
        name: "",
        contributors: [],
        version: "",
        host: [],
        apdex: ""
      });
    } else {
      setErrors(errors);
    }
  };

  return (
    <Form onSubmit={handleOnSubmit}>
      <Title>Add App</Title>
      <Label>App Name:</Label>
      <Input
        name="name"
        type="text"
        value={app.name}
        onChange={handleOnChange}
      />
      {errors.name && <Helper>{errors.name}</Helper>}
      <Label>Contributors:</Label>
      <Input
        name="contributors"
        type="text"
        value={app.contributors.toString() || ""}
        onChange={handleOnChange}
        placeholder={`Write the contributor's names separate by comma`}
      />
      {errors.contributors && <Helper>{errors.contributors}</Helper>}
      <Label>Apdex:</Label>
      <Input
        name="apdex"
        type="text"
        value={app.apdex}
        onChange={handleOnChange}
      />
      {errors.apdex && <Helper>{errors.apdex}</Helper>}
      <Label>Version:</Label>
      <Input
        name="version"
        type="text"
        value={app.version}
        onChange={handleOnChange}
      />
      {errors.version && <Helper>{errors.version}</Helper>}
      <Label>Hosts:</Label>
      <Input
        name="host"
        type="text"
        value={app.host.toString() || ""}
        onChange={handleOnChange}
        placeholder={`Write the host's names separate by comma`}
      />
      {errors.host && <Helper>{errors.host}</Helper>}
      <ButtonWrapper>
        <Button type="submit">Add</Button>
      </ButtonWrapper>
    </Form>
  );
};

AddAppForm.propTypes = {
  onAddApp: func.isRequired,
  onShowToastMessage: func.isRequired
};

export {AddAppForm};
