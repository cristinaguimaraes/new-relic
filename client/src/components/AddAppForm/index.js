import { AddAppFormContainer } from './AddAppFormContainer';
export { AddAppFormContainer as AddAppForm };