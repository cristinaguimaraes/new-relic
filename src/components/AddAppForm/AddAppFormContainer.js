import { AddAppForm } from "./AddAppForm";
import { onAddApp, onShowToastMessage } from "redux/actions/creators";
import { withRouter } from "react-router-dom";
import { connect } from "react-redux";

const mapDispatchToProps = {
  onAddApp,
  onShowToastMessage
};

export const AddAppFormContainer = withRouter(
  connect(null, mapDispatchToProps)(AddAppForm)
);
