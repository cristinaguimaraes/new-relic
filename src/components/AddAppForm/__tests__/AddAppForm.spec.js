import { AddAppForm } from "../AddAppForm";
import { shallow } from "enzyme";
import React from "react";

describe("AddAppForm component", () => {
  test("should render properly", () => {
    expect(
      shallow(
        <AddAppForm onAddApp={jest.fn()} onShowToastMessage={jest.fn()}>
          Test
        </AddAppForm>
      )
    ).toMatchSnapshot();
  });
});
