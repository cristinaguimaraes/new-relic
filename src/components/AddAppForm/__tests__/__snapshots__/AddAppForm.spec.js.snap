// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAppForm component should render properly 1`] = `
<styled.form
  onSubmit={[Function]}
>
  <styled.div>
    Add App
  </styled.div>
  <styled.label>
    App Name:
  </styled.label>
  <styled.input
    name="name"
    onChange={[Function]}
    type="text"
    value=""
  />
  <styled.label>
    Contributors:
  </styled.label>
  <styled.input
    name="contributors"
    onChange={[Function]}
    placeholder="Write the contributors names separate by comma"
    type="text"
    value=""
  />
  <styled.label>
    Apdex:
  </styled.label>
  <styled.input
    name="apdex"
    onChange={[Function]}
    type="number"
    value=""
  />
  <styled.label>
    Version:
  </styled.label>
  <styled.input
    name="version"
    onChange={[Function]}
    type="number"
    value=""
  />
  <styled.label>
    Hosts:
  </styled.label>
  <styled.input
    name="host"
    onChange={[Function]}
    placeholder="Write the hosts names separate by comma"
    type="text"
    value=""
  />
  <styled.div>
    <styled.button
      type="submit"
    >
      Add
    </styled.button>
  </styled.div>
</styled.form>
`;
